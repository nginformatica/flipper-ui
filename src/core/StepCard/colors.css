.barColorPrimary {
    background-color: #40A344!important;
}
