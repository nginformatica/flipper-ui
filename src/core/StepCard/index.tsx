/* eslint-disable max-lines */
import { Box, LinearProgress, List, ListItemText } from '@material-ui/core'
import MuiExpansionPanel from '@material-ui/core/ExpansionPanel'
import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import MuiExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary'
import { theme } from 'nginformatica-styleguide'
import React, { useMemo } from 'react'
import {
    CheckCircle as CheckCircleIcon,
    ExpandMore as ExpandMoreIcon,
    Help as HelpIcon
} from '../../icons'
import Typography from '../Typography'
import './colors.css'
import {
    ListItemContainer,
    StepCardColumn,
    StepCardRow,
    StepContainer,
    Container,
    NormalProgressContainer,
    TitleContainer
} from './styles'
import IconButton from '../IconButton'
import { PropWithDataCy } from '../types'

const { feedback, grays } = theme.colors
const DONE_COLOR = feedback.success
const UNDONE_COLOR = grays.g3
const UNDONE_FONT_COLOR = grays.g2
const UNDONE_TITLE_COLOR = grays.g1
const UNDONE_PROGRESS_COLOR = grays.g5

/**
 * Props for the StepCard component
 *
 * @interface IStepCardProps
 */
export interface IStepCardProps {
    /**
     * Whether to show the percentage on the expansion panel summary
     * Defaults to true
     * @type {boolean}
     * @memberof IStepCardProps
     * @default true
     */
    showBottomPercentage: boolean

    /**
     * A text to be displayed on the expansion panel summary
     *
     * @type {string}
     * @memberof IStepCardProps
     */
    summary: string

    /**
     * The title of the step card
     *
     * @type {string}
     * @memberof IStepCardProps
     */
    title: string

    /**
     * The subtitle of the step card title
     *
     * @type {string}
     * @memberof IStepCardProps
     */
    subTitle?: string

    /**
     * The image to be displayed on the expansion panel details
     *
     * @type {string}
     * @memberof IStepCardProps
     */
    image?: string

    /**
     * The total number of steps
     *
     * @type {number}
     * @memberof IStepCardProps
     */
    totalSteps: number

    /**
     * The number of completed steps
     *
     * @type {number}
     * @memberof IStepCardProps
     */
    doneSteps: number

    /**
     * An array of steps, each with a title, a done flag and an optional url
     *
     * @type [{
     *         title: string;
     *         done: boolean;
     *         url?: string;
     *     }]
     * @memberof IStepCardProps
     */
    steps: {
        title: string
        done: boolean
        url?: string
    }[]

    /**
     * Props to be passed to the root element
     *
     * @type {(React.HTMLAttributes<HTMLDivElement> | undefined)}
     * @memberof IStepCardProps
     */
    rootProps?: React.HTMLAttributes<HTMLDivElement>

    /**
     * Props to be passed to the title typography
     *
     * @type {(typeof Typography | undefined)}
     * @memberof IStepCardProps
     */
    titleProps?: PropWithDataCy<typeof Typography>

    /**
     * Props to be passed to the summary typography
     *
     * @type {(typeof Typography | undefined)}
     * @memberof IStepCardProps
     */
    summaryProps?: PropWithDataCy<typeof Typography>

    /**
     * Props to be passed to the expansion panel
     * @type {(typeof MuiExpansionPanel | undefined)}
     * @memberof IStepCardProps
     */
    expansionPanelDetailsProps?: PropWithDataCy<typeof MuiExpansionPanelDetails>

    /**
     * Props to be passed to the linear progress bar
     * @type {(typeof LinearProgress | undefined)}
     * @memberof IStepCardProps
     */
    linearProgressBarProps?: PropWithDataCy<typeof LinearProgress>

    /**
     * Props to be passed to the expansion panel summary linear progress bar
     * @type {(typeof LinearProgress | undefined)}
     * @memberof IStepCardProps
     */
    summaryLinearProgressBarProps?: PropWithDataCy<typeof LinearProgress>

    /**
     * Whether the expansion panel is expandable
     * @type {boolean}
     * @memberof IStepCardProps
     * @default true
     */
    expandable: boolean

    /**
     * Whether to show the icon on the expansion panel summary
     * @type {boolean}
     * @memberof IStepCardProps
     * @default true
     */
    showIcon: boolean

    /**
     * Callback to be called when the user clicks on a step url
     * @memberof IStepCardProps
     * @default undefined
     * @type {(url: string) => void}
     */
    onStepUrlClick?: (url: string) => void
}

/**
 * StepCard is a reusable component that displays a card with a title, a summary,
 * an expandable panel, and a progress bar.
 * The expandable panel displays a list of steps, each with a title, a done flag,
 * and an optional url.
 *
 * @example
 * <StepCard
 *   showBottomPercentage={true}
 *   summary="50% Complete"
 *   title="Step 1"
 *   subTitle="Subtitle for step 1"
 *   image="https://via.placeholder.com/150"
 *   totalSteps={4}
 *   doneSteps={2}
 *   steps={[
 *     { title: "Step 1-1", done: true },
 *     { title: "Step 1-2", done: true, url: "/step-1-2" },
 *     { title: "Step 1-3", done: false },
 *     { title: "Step 1-4", done: false }
 *   ]}
 *   rootProps={{ className: "my-custom-class" }}
 *   titleProps={{ variant: "h3", color: "primary" }}
 *   summaryProps={{ variant: "body2", color: "secondary" }}
 *   expandable={true}
 *   defaultExpanded={false}
 * />
 */

const StepCard = (props: IStepCardProps) => {
    const {
        steps,
        title,
        titleProps,
        subTitle,
        image,
        showBottomPercentage = true,
        expandable = true,
        showIcon = true,
        doneSteps,
        totalSteps,
        summary,
        summaryProps,
        expansionPanelDetailsProps,
        linearProgressBarProps,
        rootProps,
        onStepUrlClick
    } = props
    const percentage = Math.round((doneSteps / totalSteps) * 100)

    const renderTitleColumn = useMemo(
        () => (
            <>
                {showIcon && (
                    <CheckCircleIcon
                        style={{
                            fontSize: 40,
                            color:
                                percentage === 100 ? DONE_COLOR : UNDONE_COLOR
                        }}
                    />
                )}
                <TitleContainer>
                    <Typography
                        variant='h6'
                        style={{
                            color: UNDONE_TITLE_COLOR,
                            fontSize: 24,
                            textAlign: 'center'
                        }}
                        {...titleProps}>
                        {title}
                    </Typography>
                    {subTitle && (
                        <Typography
                            variant='h6'
                            style={{
                                color: UNDONE_COLOR,
                                textAlign: 'center'
                            }}
                            {...titleProps}>
                            {subTitle}
                        </Typography>
                    )}
                </TitleContainer>
            </>
        ),
        [showIcon, percentage, titleProps, title, subTitle]
    )

    const renderSummaryColumn = useMemo(
        () => (
            <>
                <MuiExpansionPanelSummary
                    style={{ textAlign: 'center' }}
                    expandIcon={<ExpandMoreIcon fontSize='large' />}>
                    <Typography variant='h6' {...summaryProps}>
                        {summary}
                    </Typography>
                </MuiExpansionPanelSummary>
            </>
        ),
        [summary, summaryProps]
    )

    const renderStepsList = useMemo(
        () => (
            <List>
                {steps.map((step, index) => {
                    const handleOnClickStepUrl = () => {
                        if (onStepUrlClick && step.url) {
                            onStepUrlClick(step.url)
                        }
                    }

                    return (
                        <ListItemText
                            key={index}
                            primary={
                                <ListItemContainer>
                                    <CheckCircleIcon
                                        style={{
                                            color: step.done
                                                ? DONE_COLOR
                                                : UNDONE_COLOR
                                        }}
                                    />
                                    <Typography
                                        variant='body1'
                                        style={{
                                            color: UNDONE_FONT_COLOR,
                                            fontWeight: 600
                                        }}>
                                        {step.title}
                                    </Typography>
                                    {step.url && onStepUrlClick && (
                                        <>
                                            <IconButton
                                                style={{
                                                    backgroundColor: 'unset'
                                                }}
                                                onClick={handleOnClickStepUrl}>
                                                <HelpIcon
                                                    fontSize='small'
                                                    style={{ margin: '0px' }}
                                                />
                                            </IconButton>
                                        </>
                                    )}
                                </ListItemContainer>
                            }
                        />
                    )
                })}
            </List>
        ),
        [onStepUrlClick, steps]
    )

    const renderNormalProgress = useMemo(
        () => (
            <NormalProgressContainer>
                <Typography variant='h6' {...summaryProps}>
                    {summary}
                </Typography>
                <Typography
                    variant='subtitle2'
                    style={{
                        position: 'absolute',
                        right: 8
                    }}>{`${percentage}%`}</Typography>
                <Box style={{ width: '100%' }}>
                    <LinearProgress
                        variant='determinate'
                        value={percentage}
                        color='primary'
                        classes={{
                            barColorPrimary: 'barColorPrimary'
                        }}
                        style={{
                            borderRadius: 10,
                            height: '16px',
                            backgroundColor: UNDONE_PROGRESS_COLOR
                        }}
                    />
                </Box>
            </NormalProgressContainer>
        ),
        [percentage, summary, summaryProps]
    )

    return (
        <Container {...rootProps}>
            <MuiExpansionPanel>
                <StepContainer>
                    <StepCardRow height={expandable ? '100px' : '100%'}>
                        <StepCardColumn justifyContent='start'>
                            {renderTitleColumn}
                        </StepCardColumn>
                        <StepCardColumn justifyContent='end'>
                            {expandable
                                ? renderSummaryColumn
                                : renderNormalProgress}
                        </StepCardColumn>
                    </StepCardRow>
                    {showBottomPercentage && (
                        <StepCardRow height='15px'>
                            <Box style={{ width: '100%' }}>
                                <LinearProgress
                                    variant='determinate'
                                    value={percentage}
                                    color='primary'
                                    classes={{
                                        barColorPrimary: 'barColorPrimary'
                                    }}
                                    style={{
                                        height: '16px',
                                        backgroundColor: UNDONE_PROGRESS_COLOR
                                    }}
                                    {...linearProgressBarProps}
                                />
                            </Box>
                        </StepCardRow>
                    )}
                </StepContainer>
                {expandable && (
                    <MuiExpansionPanelDetails
                        style={{
                            display: 'flex',
                            justifyContent: 'space-between'
                        }}
                        {...expansionPanelDetailsProps}>
                        <StepCardColumn justifyContent='start'>
                            {renderStepsList}
                        </StepCardColumn>
                        {image && (
                            <StepCardColumn justifyContent='end'>
                                <img src={image} alt='Step Icon' />
                            </StepCardColumn>
                        )}
                    </MuiExpansionPanelDetails>
                )}
            </MuiExpansionPanel>
        </Container>
    )
}

export default StepCard
