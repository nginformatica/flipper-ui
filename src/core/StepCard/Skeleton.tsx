import { Box } from '@material-ui/core'
import MuiExpansionPanel from '@material-ui/core/ExpansionPanel'
import { Skeleton } from '@material-ui/lab'
import React from 'react'
import './colors.css'
import {
    Container,
    NormalProgressContainer,
    StepCardColumn,
    StepCardRow,
    StepContainer,
    TitleContainer
} from './styles'

interface IStepCardPanelProps {
    expandable: boolean
    showBottomPercentage: boolean
    showIcon: boolean
    subTitle?: boolean
}

const StepCardSkeleton = (props: IStepCardPanelProps) => {
    const { expandable, showBottomPercentage, showIcon, subTitle } = props

    const TextSkeleton = () => (
        <Skeleton variant='text' style={{ minWidth: '400px' }} height={40} />
    )

    const NormalProgress = () => (
        <NormalProgressContainer>
            <Skeleton
                variant='text'
                height={40}
                style={{ minWidth: '100px', margin: '4px 0px' }}
            />
            <Box style={{ width: '100%' }}>
                <Skeleton variant='text' />
            </Box>
        </NormalProgressContainer>
    )

    return (
        <Container>
            <MuiExpansionPanel>
                <StepContainer withPadding={!expandable}>
                    <StepCardRow height={expandable ? '100px' : '100%'}>
                        <StepCardColumn justifyContent='start'>
                            {showIcon && (
                                <Skeleton
                                    variant='circle'
                                    width={40}
                                    height={40}
                                />
                            )}
                            <TitleContainer>
                                <TextSkeleton />
                                {subTitle && <TextSkeleton />}
                            </TitleContainer>
                        </StepCardColumn>
                        <StepCardColumn justifyContent='end'>
                            {expandable ? <TextSkeleton /> : <NormalProgress />}
                        </StepCardColumn>
                    </StepCardRow>
                    {showBottomPercentage && (
                        <StepCardRow height='15px'>
                            <Box style={{ width: '100%' }}>
                                <Skeleton variant='text' />
                            </Box>
                        </StepCardRow>
                    )}
                </StepContainer>
            </MuiExpansionPanel>
        </Container>
    )
}

export default StepCardSkeleton
