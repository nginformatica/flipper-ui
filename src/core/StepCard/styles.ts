import { Box } from '@material-ui/core'
import { theme } from 'nginformatica-styleguide'
import styled from 'styled-components'

interface IStepContainerProps {
    padding?: React.CSSProperties['padding']
}

interface IRowProps {
    minHeight: React.CSSProperties['minHeight']
    fullWidth?: boolean
}

interface IContainerProps {
    fullWidth?: boolean
    margin?: React.CSSProperties['margin']
}

interface IColumnProps {
    justifyContent: React.CSSProperties['justifyContent']
}

export const Container = styled.div<IContainerProps>`
    width: 100%;
    max-width: ${props => (props.fullWidth ? '100%' : '1100px')};
    ${props => (props.margin ? `margin: ${props.margin};` : '')}
`

export const StepCardRow = styled.div<IRowProps>`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 100%;
    max-width: ${props => (props.fullWidth ? '100%' : '1100px')};
    min-height: ${props => props.minHeight};
`

export const StepContainer = styled.div<IStepContainerProps>`
    display: flex;
    flex-direction: column;
    ${props => (props.padding ? `padding: ${props.padding};` : '')}
`

export const StepCardColumn = styled.div<IColumnProps>`
    padding-inline: 30px;
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: ${props => props.justifyContent};
    width: 100%;

    ${props => {
        if (props.justifyContent === 'start') {
            return `
                svg:first-child {
                    margin-right: 40px;
                }
                `
        } else {
            return ''
        }
    }}
`

export const ListItemContainer = styled.p`
    display: flex;
    flex-direction: row;
    align-items: center;
    height: 30px;

    svg:first-child {
        margin-right: 16px;
    }
`

export const TitleContainer = styled.div`
    h6:first-child {
        padding-inline: 20px;
    }

    span:first-child {
        margin-left: 20px;
    }
`

export const NormalProgressContainer = styled.div`
    width: 100%;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    position: relative;

    h6:first-child {
        margin-bottom: 8px;
    }
`

export const BarWrapper = styled(Box)`
    width: 100%;

    & .barColorPrimary {
        background-color: ${theme.colors.feedback.success};
    }
`
