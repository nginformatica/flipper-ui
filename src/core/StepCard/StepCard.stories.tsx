import React, { useState } from 'react'
import { Meta, StoryFn } from '@storybook/react'
import StepCard from '.'
import Button from '../Button'
import { stepCardImageUrl } from '../../lib/images'

export default {
    title: 'StepCard',
    component: StepCard
} as Meta<typeof StepCard>

const Template: StoryFn<typeof StepCard> = args => {
    return <StepCard {...args} />
}

const TemplateWithLoading: StoryFn<typeof StepCard> = args => {
    const [loading, setLoading] = useState(true)

    setTimeout(() => {
        setLoading(false)
    }, 22500)

    return (
        <>
            <Button />
            <StepCard {...args} loading={loading} subTitle='asd' />
        </>
    )
}

export const Default = Template.bind({})

Default.args = {
    summary: '%i passos restantes - Leva %i minuto(s)',
    time: 2,
    percentage: 75,
    remainingSteps: 3,
    title: 'Configurar a organização e filial',
    image: stepCardImageUrl,
    onStepUrlClick: (url: string) => alert(url),
    steps: [
        {
            title: 'Informar Razão social',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados do eSocial',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados da filial(CNAE, CNPJ)',
            done: false,
            url: 'https://www.google.com'
        }
    ]
}

export const WithoutExpandable = Template.bind({})

WithoutExpandable.args = {
    loading: false,
    padding: '30px',
    percentage: 75,
    showBottomPercentage: false,
    expandable: false,
    showIcon: false,
    title: 'Conheça mais sobre o Quírons',
    subTitle:
        'Siga os passos recomendados para uma melhor experiência ' +
        'dentro da plataforma.',
    summary: 'Seu progresso',
    steps: [
        {
            title: 'Informar Razão social',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados do eSocial',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados da filial(CNAE, CNPJ)',
            done: false,
            url: 'https://www.google.com'
        }
    ]
}
export const WithLoading = TemplateWithLoading.bind({})

WithLoading.args = {
    summary: '%i passos restantes - Leva %i minuto(s)',
    time: 2,
    remainingSteps: 3,
    title: 'Configurar a organização e filial',
    image: stepCardImageUrl,
    onStepUrlClick: (url: string) => alert(url),
    steps: [
        {
            title: 'Informar Razão social',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados do eSocial',
            done: true,
            url: 'https://www.google.com'
        },
        {
            title: 'Informar dados da filial(CNAE, CNPJ)',
            done: false,
            url: 'https://www.google.com'
        }
    ]
}
