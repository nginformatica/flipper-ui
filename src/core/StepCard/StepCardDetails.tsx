import { List, ListItemText } from '@material-ui/core'
import MuiExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails'
import { theme } from 'nginformatica-styleguide'
import React from 'react'
import { IStepCardProps } from '.'
import { CheckCircle as CheckCircleIcon, Help as HelpIcon } from '../../icons'
import IconButton from '../IconButton'
import Typography from '../Typography'
import { ListItemContainer, StepCardColumn } from './styles'

const { feedback, grays } = theme.colors
const DONE_COLOR = feedback.success
const UNDONE_COLOR = grays.g3
const UNDONE_FONT_COLOR = grays.g2

interface IStepCardDetailsProps {
    steps: IStepCardProps['steps']
    onStepUrlClick?: IStepCardProps['onStepUrlClick']
    image?: IStepCardProps['image']
    expansionPanelDetailsProps?: IStepCardProps['expansionPanelDetailsProps']
}

export const StepCardDetails = (props: IStepCardDetailsProps) => {
    const { steps, onStepUrlClick, image, expansionPanelDetailsProps } = props

    const StepsList = () => (
        <List>
            {(steps || []).map((step, index) => {
                const handleOnClickStepUrl = () => {
                    if (onStepUrlClick && step.url) {
                        onStepUrlClick(step.url)
                    }
                }

                return (
                    <ListItemText
                        key={index}
                        primary={
                            <ListItemContainer>
                                <CheckCircleIcon
                                    style={{
                                        color: step.done
                                            ? DONE_COLOR
                                            : UNDONE_COLOR
                                    }}
                                />
                                <Typography
                                    variant='body1'
                                    style={{
                                        color: UNDONE_FONT_COLOR,
                                        fontWeight: 600
                                    }}>
                                    {step.title}
                                </Typography>
                                {step.url && onStepUrlClick && (
                                    <>
                                        <IconButton
                                            style={{
                                                backgroundColor: 'unset'
                                            }}
                                            onClick={handleOnClickStepUrl}>
                                            <HelpIcon
                                                fontSize='small'
                                                style={{ margin: '0px' }}
                                            />
                                        </IconButton>
                                    </>
                                )}
                            </ListItemContainer>
                        }
                    />
                )
            })}
        </List>
    )

    return (
        <MuiExpansionPanelDetails
            style={{
                display: 'flex',
                justifyContent: 'space-between'
            }}
            {...expansionPanelDetailsProps}>
            <StepCardColumn justifyContent='start'>
                <StepsList />
            </StepCardColumn>
            {image && (
                <StepCardColumn justifyContent='end'>
                    {typeof image === 'string' ? (
                        <img src={image} alt='Step Icon' />
                    ) : (
                        image
                    )}
                </StepCardColumn>
            )}
        </MuiExpansionPanelDetails>
    )
}
