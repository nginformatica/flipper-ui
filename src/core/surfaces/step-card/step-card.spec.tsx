import React from 'react'
import { render, screen, waitFor } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event/'
import StepCard from '@/test/mocks/step-card-mock'

describe('StepCard', () => {
    it('should render', () => {
        render(
            <StepCard
                stepProps={{
                    percentage: 0,
                    summary: 'Summary',
                    subTitle: 'Subtitle',
                    title: 'Title'
                }}
            />
        )

        const title = screen.getByText('Title')
        const summary = screen.getByText('Summary')
        const subTitle = screen.getByText('Subtitle')
        const statusIcon = screen.getByTestId('step-card-title-icon')

        expect(title).toBeDefined()
        expect(summary).toBeDefined()
        expect(subTitle).toBeDefined()
        expect(statusIcon.style.color).toBe('rgb(117, 117, 117)')
    })

    it('should render with custom styles', () => {
        render(
            <StepCard
                stepProps={{
                    margin: '10px',
                    padding: '5px',
                    percentage: 0,
                    summary: 'Summary',
                    subTitle: 'Subtitle',
                    title: 'Title'
                }}
            />
        )

        const title = screen.getByText('Title')
        const summary = screen.getByText('Summary')
        const subTitle = screen.getByText('Subtitle')

        expect(title).toBeDefined()
        expect(summary).toBeDefined()
        expect(subTitle).toBeDefined()
    })

    it('should render with no loading state', () => {
        render(
            <StepCard
                blockInitialLoading
                stepProps={{
                    expanded: true,
                    percentage: 0,
                    summary: 'Summary',
                    subTitle: 'Subtitle',
                    title: 'Title'
                }}
            />
        )

        const title = screen.getByText('Title')
        const summary = screen.getByText('Summary')
        const subTitle = screen.getByText('Subtitle')

        expect(title).toBeDefined()
        expect(summary).toBeDefined()
        expect(subTitle).toBeDefined()
    })

    it('should render with loading', async () => {
        jest.useFakeTimers()

        render(
            <StepCard
                initialLoading
                stepProps={{
                    percentage: 0,
                    summary: 'Summary',
                    subTitle: 'Subtitle',
                    title: 'Title'
                }}
            />
        )
        const skeleton = await screen.findByTestId('skeleton-container')

        expect(skeleton).toBeDefined()

        jest.runAllTimers()

        await waitFor(async () => {
            const title = screen.getByText('Title')
            const summary = screen.getByText('Summary')
            const subTitle = screen.getByText('Subtitle')

            expect(title).toBeDefined()
            expect(summary).toBeDefined()
            expect(subTitle).toBeDefined()
        })

        jest.useRealTimers()
    })

    it('should render without expandable but with loading', async () => {
        jest.useFakeTimers()

        render(
            <StepCard
                initialLoading
                stepProps={{
                    percentage: 0,
                    showSubTitleSkeleton: true,
                    expandable: false,
                    summary: 'Summary',
                    subTitle: 'Subtitle',
                    title: 'Title'
                }}
            />
        )
        const skeleton = await screen.findByTestId('skeleton-container')

        expect(skeleton).toBeDefined()

        jest.runAllTimers()

        await waitFor(async () => {
            const title = screen.getByText('Title')
            const summary = screen.getByText('Summary')
            const subTitle = screen.queryByText('Subtitle')

            expect(title).toBeDefined()
            expect(summary).toBeDefined()
            expect(subTitle).toBeDefined()
        })

        jest.useRealTimers()
    })

    it('should render without expandable', () => {
        render(
            <StepCard
                stepProps={{
                    percentage: 0,
                    expandable: false,
                    fullWidth: true,
                    summary: 'Summary',
                    title: 'Title'
                }}
            />
        )

        const progressBar = screen.getByTestId('normal-progress')

        expect(progressBar.lastElementChild?.className).toContain(
            'MuiLinearProgress-root'
        )
    })

    it('should render step 100% completed', () => {
        render(
            <StepCard
                stepProps={{
                    percentage: 100,
                    summary: 'Summary',
                    title: 'Title'
                }}
            />
        )

        const statusIcon = screen.getByTestId('step-card-title-icon')

        expect(statusIcon.style.color).toBe('rgb(40, 153, 57)')
    })

    it('should expand without image', async () => {
        render(
            <StepCard
                stepProps={{
                    percentage: 0,
                    summary: 'Summary',
                    title: 'Title',
                    steps: [
                        {
                            title: 'Step 1',
                            done: false
                        },
                        {
                            title: 'Step 2',
                            done: true
                        }
                    ]
                }}
            />
        )

        const title = screen.getByText('Title')

        await userEvent.click(title)

        const step1 = screen.getByText('Step 1')
        const step2 = screen.getByText('Step 2')

        expect(step1).toBeDefined()
        expect(step2).toBeDefined()
    })

    it('should expand with image as string', async () => {
        render(
            <StepCard
                stepProps={{
                    image: 'https://picsum.photos/200',
                    percentage: 50,
                    summary: 'Summary',
                    title: 'Title',
                    steps: [
                        {
                            title: 'Step 1',
                            done: false
                        },
                        {
                            title: 'Step 2',
                            done: true
                        }
                    ]
                }}
            />
        )

        const title = screen.getByText('Title')

        await userEvent.click(title)

        const thumbnail = screen.getByAltText('Step Icon')

        expect(thumbnail).toBeDefined()
    })

    it('should expand with image as object', async () => {
        render(
            <StepCard
                stepProps={{
                    image: <img src='valid-url' alt='Step Icon' />,
                    percentage: 50,
                    summary: 'Summary',
                    title: 'Title',
                    steps: [
                        {
                            title: 'Step 1',
                            done: false
                        },
                        {
                            title: 'Step 2',
                            done: true
                        }
                    ]
                }}
            />
        )

        const title = screen.getByText('Title')

        await userEvent.click(title)

        const thumbnail = screen.getByAltText('Step Icon')

        expect(thumbnail).toBeDefined()
    })

    it('should call onStepUrlClick with correct values', async () => {
        const onStepUrlClickSpy = jest.fn()

        render(
            <StepCard
                stepProps={{
                    percentage: 0,
                    summary: 'Summary',
                    title: 'Title',
                    onStepUrlClick: onStepUrlClickSpy,
                    steps: [
                        {
                            title: 'Step 1',
                            done: false,
                            url: 'valid-url'
                        },
                        {
                            title: 'Step 2',
                            done: true
                        }
                    ]
                }}
            />
        )

        const title = screen.getByText('Title')

        await userEvent.click(title)

        const step1InfoButton = screen.getByTestId('step-card-button-0')

        await userEvent.click(step1InfoButton)

        expect(onStepUrlClickSpy).toHaveBeenCalledWith('valid-url')
    })

    it('should match snapshot', () => {
        const { container } = render(
            <StepCard
                stepProps={{
                    image: 'https://picsum.photos/200',
                    percentage: 50,
                    summary: 'Summary',
                    title: 'Title',
                    steps: [
                        {
                            title: 'Step 1',
                            done: false
                        },
                        {
                            title: 'Step 2',
                            done: true
                        }
                    ]
                }}
            />
        )

        expect(container).toMatchSnapshot()
    })
})
