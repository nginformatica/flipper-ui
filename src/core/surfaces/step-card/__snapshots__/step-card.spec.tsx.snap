// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepCard should match snapshot 1`] = `
<div>
  <div
    class="sc-aYaIB hvFdbM"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1086bdv-MuiPaper-root-MuiAccordion-root"
    >
      <div
        class="sc-eqUzNf jwXvAb"
      >
        <div
          class="sc-gEvDqW dsflTG"
          minheight="100px"
        >
          <div
            class="sc-fqkwJk ecodSZ"
            justifycontent="start"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="step-card-title-icon"
              focusable="false"
              style="font-size: 40px; color: rgb(117, 117, 117);"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
              />
            </svg>
            <div
              class="sc-iGgVNO dRwqyP"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6"
                style="color: rgb(66, 66, 66); font-size: 24px; text-align: center;"
              >
                Title
              </h6>
            </div>
          </div>
          <div
            class="sc-fqkwJk iZCAOH"
            justifycontent="end"
          >
            <div
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-1f773le-MuiButtonBase-root-MuiAccordionSummary-root"
              role="button"
              style="text-align: center;"
              tabindex="0"
            >
              <div
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-eqpfi5-MuiAccordionSummary-content"
              >
                <h6
                  class="MuiTypography-root MuiTypography-h6"
                >
                  Summary
                </h6>
              </div>
              <div
                class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-gEvDqW eCImaH"
          minheight="15px"
        >
          <span
            aria-valuemax="100"
            aria-valuemin="0"
            aria-valuenow="50"
            class="MuiLinearProgress-root MuiLinearProgress-colorSuccess MuiLinearProgress-determinate css-1hxt08m-MuiLinearProgress-root"
            role="progressbar"
            style="height: 16px; width: 100%; background-color: rgb(189, 189, 189);"
          >
            <span
              class="MuiLinearProgress-bar MuiLinearProgress-barColorSuccess MuiLinearProgress-bar1Determinate css-h8g9t7-MuiLinearProgress-bar1"
              style="transform: translateX(-50%);"
            />
          </span>
        </div>
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiAccordion-region"
              role="region"
            >
              <div
                class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  class="sc-fqkwJk ecodSZ"
                  justifycontent="start"
                >
                  <ul
                    class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
                  >
                    <li
                      class="sc-dcJtft jbcafN"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckCircleIcon"
                        focusable="false"
                        style="color: rgb(117, 117, 117);"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
                        />
                      </svg>
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                        style="color: rgb(97, 97, 97); font-weight: 600;"
                      >
                        Step 1
                      </p>
                    </li>
                    <li
                      class="sc-dcJtft jbcafN"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckCircleIcon"
                        focusable="false"
                        style="color: rgb(14, 144, 67);"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"
                        />
                      </svg>
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                        style="color: rgb(97, 97, 97); font-weight: 600;"
                      >
                        Step 2
                      </p>
                    </li>
                  </ul>
                </div>
                <div
                  class="sc-fqkwJk iZCAOH"
                  justifycontent="end"
                >
                  <img
                    alt="Step Icon"
                    src="https://picsum.photos/200"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
